### PR TITLE
Multi set structure

### PR DESCRIPTION
**Pull request by mzeuner**
We introduce a simple signature for multi-set structures in the spirit of Okasaki's signature for sets (p. 12 of "Purely Functional Data Structures"). We then define multi-set structures on finite multi-sets and association lists and show that the two are isomorphic.

### CODE REVIEW

**Comment by mortberg** [comment]
Nice! Ping @cangiuli

**Comment by mortberg** [comment]
As you're now using the recursor/eliminator and not pattern-matching, shouldn't `with` work so that the helper definitions can be inlined?

File: Cubical/HITs/AssocList/Properties.agda [hunk 1]
```diff
@@ -72,6 +76,108 @@ FMS→AL∘AL→FMS≡id = AL.ElimProp.f (AL.trunc _ _) refl (λ x n {xs} p →
...
+ ALmember a = AL.Rec.f isSetℕ 0 (ALmember-⟨,⟩∷* a) (ALmember-per* a) (ALmember-agg* a) (ALmember-del* a)
+
+
+
+ AL-with-str : Multi-Set A (Discrete→isSet discA)
+ AL-with-str = (AssocList A , ⟨⟩ , ⟨_, 1 ⟩∷_ , ALmember)
+
+
+-- We want to show that Al-with-str ≅ FMS-with-str as multiset-structures
+module _(discA : Discrete A) where
```
Comment (mortberg) [inline Cubical/HITs/AssocList/Properties.agda:149]: Why make a new module here and just repeat the discA assumption?

File: Cubical/HITs/AssocList/Properties.agda [hunk 2]
```diff
@@ -72,6 +76,108 @@ FMS→AL∘AL→FMS≡id = AL.ElimProp.f (AL.trunc _ _) refl (λ x n {xs} p →
...
+    α : ⟨ x , 1 ⟩∷ FMS→AL xs ≡ FMS→AL (x ∷ xs)
+    α = sym (multi-∷-id x 1 xs)
+
+    θ : FMSmember discA a (x ∷ xs) ≡ ALmember discA a (⟨ x , 1 ⟩∷ (FMS→AL xs))
+    θ = χ x (FMSmember discA a xs) (ALmember discA a (FMS→AL xs)) (discA a x) p
+
+
+
+ FMS-with-str≡AL-with-str : FMS-with-str discA ≡ AL-with-str discA
+ FMS-with-str≡AL-with-str = SIP (multi-set-structure A (Discrete→isSet discA))
```
Comment (mortberg) [inline Cubical/HITs/AssocList/Properties.agda:178]: Instead of calling `Discrete→isSet discA` repeatedly just make it a private definition in the model.

Comment (mortberg) [reply Cubical/HITs/AssocList/Properties.agda:178]: module :-)

File: Cubical/HITs/AssocList/Properties.agda (hunk 1 above)
Comment (mzeuner) [reply Cubical/HITs/AssocList/Properties.agda:149]: good question, I don't remember why I did that, maybe wanted to do something in between initially...

**Comment by mzeuner** [comment]
Some of the aux-function are needed  in later proofs, see e.g. 
χ :  (x : A) (xs ys : ℕ) (p : Dec (a ≡ x))
      → xs ≡ ys
      → FMSmember-∷*-aux a x p xs ≡ ALmember-⟨,⟩∷*-aux a x p 1 ys
in the proof of FMS→AL-isIso, I guess I could drop all the other aux-function and do it with with, but maybe they will be needed later as well...

**Comment by mortberg** [comment]
> Some of the aux-function are needed in later proofs, see e.g.
> χ : (x : A) (xs ys : ℕ) (p : Dec (a ≡ x))
> → xs ≡ ys
> → FMSmember-∷*-aux a x p xs ≡ ALmember-⟨,⟩∷*-aux a x p 1 ys
> in the proof of FMS→AL-isIso, I guess I could drop all the other aux-function and do it with with, but maybe they will be needed later as well...

Ah I see! I guess it's better to have them like this then.

Merging now.

**Pull request merged**